xbmc/addons/AddonManager.cpp

```diff
@@ -297,29 +298,90 @@ bool CAddonMgr::HasAvailableUpdates()
   return !GetAvailableUpdates().empty();
 }
 
+std::vector<std::shared_ptr<IAddon>> CAddonMgr::GetOrphanedDependencies() const
+{
+  std::vector<std::shared_ptr<IAddon>> allAddons;
+  GetAddonsInternal(ADDON_UNKNOWN, allAddons, OnlyEnabled::YES, CheckIncompatible::YES);
+
+  std::vector<std::shared_ptr<IAddon>> orphanedDependencies;
+  for (const auto& addon : allAddons)
+  {
+    if (IsOrphaned(addon, allAddons))
+    {
+      orphanedDependencies.emplace_back(addon);
+    }
+  }
+
+  return orphanedDependencies;
+}
+
+bool CAddonMgr::IsOrphaned(const std::shared_ptr<IAddon>& addon,
+                           const std::vector<std::shared_ptr<IAddon>>& allAddons) const
+{
+  if (CServiceBroker::GetAddonMgr().IsSystemAddon(addon->ID()) ||
+      !CAddonType::IsDependencyType(addon->MainType()))
+    return false;
+
+  for (const auto& other : allAddons)
+  {
+    const auto& deps = other->GetDependencies();
+    auto it = std::find_if(deps.begin(), deps.end(),
+                           [&](const DependencyInfo& dep) { return dep.id == addon->ID(); });
+    if (it != deps.end())
+      return false;
+  }
+  return true;
+}
+
+std::vector<std::string> CAddonMgr::RemoveOrphanedDependencies() const
+{
+  std::vector<std::string> removedItems;
+
+  for (auto toRemove = GetOrphanedDependencies(); toRemove.size() > 0;
+       toRemove = GetOrphanedDependencies())
```

Comment:
This looks strange. Could this be expressed shorter/more elegant with a `while` loop?

Reply:
i had a `do...while` in the first place but still needed to initialize `toRemove`. let's see if i find something prettier

Reply:
```cpp
auto toRemove = GetOrphanedDependencies();
while (toRemove.size() > 0)
  toRemove = GetOrphanedDependencies())
```

Reply:
thanks, done